ec/acc/deployment_memory_optimized_test.go

```diff
@@ -0,0 +1,95 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package acc
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccDeployment_memoryOptimized(t *testing.T) {
```

Comment:
Am I right in assuming that the tests in general run some HCL config and then run some assertions on resulting state ?
If so would it be worth to come up with a way to match that against an expected json with a nice message on how they differ if different ? That would make the tests more readable, and we could even set them up in a series of HCP -> state json ( or yml ) without requiring any code. I think that would lower the barrier for adding tests, e.g. it would be easy for someone to contribute a test that reproeduces the issue with an issue report.

Reply:
Yes! the configs that are used to perform CRUD actions can be found in [ec/acc/testdata/](https://github.com/elastic/terraform-provider-ec/tree/master/ec/acc/testdata).

Fortunately, what you are requesting is already done by the terraform-plugin-sdk's [testing framework](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.4/helper/resource#TestCheckFunc). The framework has several functions like `resource.TestCheckResourceAttrSet` which checks whether an attribute is set or not, `resource.TestCheckResourceAttr` which compares an attribute with what is expected, `resource.TestCheckResourceAttrPair` which compares two different attributes to see if they match and several others.

This means that, for example, when we use the `resource.TestCheckResourceAttrPair` function and the two attributes don't match we get this message.

```console
Check failed: Check 7/10 error: data.ec_deployments.query: Attribute "deployments.0.elasticsearch_resource_id" not set, but "elasticsearch.0.resource_id" is set in ec_deployment.basic_datasource as "008d32ee11fc4c3b9ed104062f0623d2"
```